### PR TITLE
M975

### PR DESCRIPTION
**Pull request by gwan93**
For https://trello.com/c/GmwshTS5/975-make-a-checkbox-that-shows-projects-with-0-surveys-in-the-project-table?filter=label:dash%20app

- Add a checkbox to show projects with 0 surveys
- When the table is unmounted, it resets the `showProjectsWithNoRecords` back to false
- Added some keys (countries, organizations, years) to each project object. Previously those read from the first record of each project, but if a project was filtered down to 0 records, they wouldn't be readable. 
- Changed the order of .map() and .filter(). The .map()s filter out records (projects mapped, records filtered) which would cause some records to be filtered out too early
- Some logic in the table updated to read from the newly created countries/organizations/years keys

### CODE REVIEW

**Review by alanjleonard — CHANGES_REQUESTED** [review]
The whole button should have a hover state and toggle the checkbox. Use `cursor: pointer`.
Checking and unchecking the box, shouldn't clear how the table is sorted.

**Comment by gwan93** [comment]
[Make a checkbox that shows projects with 0 surveys in the project table.](https://trello.com/c/GmwshTS5/975-make-a-checkbox-that-shows-projects-with-0-surveys-in-the-project-table)

**Review by alanjleonard — APPROVED** [review]

File: src/context/FilterProjectsContext.jsx [hunk 1]
```diff
@@ -232,155 +233,179 @@ export const FilterProjectsProvider = ({ children }) => {
...
-              const sampleUnitExists = record.protocols[name]?.sample_unit_count !== undefined
-              const isPolicyValueMatch = record[policy] === value
-              return sampleUnitExists && isPolicyValueMatch
-            })
-            // Return the inverse of shouldOmitRecord to keep only the records that should not be omitted
-            return !shouldOmitRecord
+      return (
+        projectData.results
+          // Create new keys at the project level to store the countries, organizations, and years
+          // These will persist even if the project has no records
```
Comment (julia-rom) [inline src/context/FilterProjectsContext.jsx:239]: is there plans to use this commented out code? can we remove it?

File: src/components/MapAndTableControls/components/ViewToggle.jsx [hunk 2]
```diff
@@ -119,6 +143,20 @@ const ViewToggle = ({ view, setView }) => {
...
             ref={csvLinkRef}
           />
+          {selectedOrganizations.length || showYourData ? (
+            <ZeroSurveysButton onClick={handleShowProjectsWithNoRecords}>
+              <input
+                type="checkbox"
+                id="show-0-projects"
+                checked={showProjectsWithNoRecords}
+                onChange={handleShowProjectsWithNoRecords}
+                onClick={(e) => e.stopPropagation()}
```
Comment (julia-rom) [inline src/components/MapAndTableControls/components/ViewToggle.jsx:153]: is `(e) => e.stopPropagation()` necessary on input and label?

File: src/context/FilterProjectsContext.jsx [hunk 3]
```diff
@@ -232,155 +233,179 @@ export const FilterProjectsProvider = ({ children }) => {
...
+          .filter((project) => {
+            // Filter by selected countries
+            const matchesSelectedCountries =
+              selectedCountries.length === 0 ||
+              selectedCountries.includes(project.records[0]?.country_name)
+
+            // Filter by selected organizations
+            const matchesSelectedOrganizations =
+              selectedOrganizations.length === 0 ||
+              // project.records[0]?.tags?.some((tag) => selectedOrganizations.includes(tag.name))
```
Comment (julia-rom) [inline src/context/FilterProjectsContext.jsx:265]: same for this commented out code

File: src/context/FilterProjectsContext.jsx [hunk 4]
```diff
@@ -232,155 +233,179 @@ export const FilterProjectsProvider = ({ children }) => {
...
+              ...project,
+              records: project.records.filter((record) => {
+                const recordDate = new Date(record.sample_date)
+                return recordDate <= new Date(finishDate) && recordDate >= new Date(beginDate)
+              }),
+            }
           })
+          .map((project) => {
+            // Applies method data sharing filters
+            const policyMappings = {
```
Comment (julia-rom) [inline src/context/FilterProjectsContext.jsx:305]: can we move `policyMappings` outside of code block maybe in the language file?

**Review by julia-rom — CHANGES_REQUESTED** [review]
left a few suggestions for commented out code and moving large `consts`/language mappings to an external file

File: src/context/FilterProjectsContext.jsx [hunk 5]
```diff
@@ -233,179 +233,87 @@ export const FilterProjectsProvider = ({ children }) => {
...
+              const recordDate = new Date(record.sample_date)
+              return recordDate <= new Date(finishDate) && recordDate >= new Date(beginDate)
+            }),
+          }
+        })
+        .map((project) => {
+          // Applies method data sharing filters
+          const filteredRecords = project.records.filter((record) => {
+            // Check if this record should be excluded based on methodDataSharingFilters
+            const shouldOmitRecord = methodDataSharingFilters.some((filter) => {
```
Comment (julia-rom) [inline src/context/FilterProjectsContext.jsx:285]: does it make sense to name this to `omitRecord`

File: src/context/FilterProjectsContext.jsx [hunk 6]
```diff
@@ -233,179 +233,87 @@ export const FilterProjectsProvider = ({ children }) => {
...
+              const isPolicyValueMatch = record[policy] === value
+              return sampleUnitExists && isPolicyValueMatch
             })
-
-            return {
-              ...project,
-              records: filteredRecords,
-            }
+            // Return the inverse of shouldOmitRecord to keep only the records that should not be omitted
+            return !shouldOmitRecord
```
Comment (julia-rom) [inline src/context/FilterProjectsContext.jsx:292]: this is confusing .. would prefer a more descriptive var name

Comment (julia-rom) [reply src/context/FilterProjectsContext.jsx:292]: `omitRecordIf` (exact reason the record should be omitted)

**Review by julia-rom — COMMENTED** [review]
looking better, would like to see a more descriptive var name for `shouldOmitRecord`

**Comment by mmnoo** [comment]
[Update tool-bar layout](https://trello.com/c/DwZux8Jf/1029-update-tool-bar-layout)

**Pull request merged**